sdk/typescript/src/types/framework.ts

```diff
@@ -57,18 +62,109 @@ export class Coin {
     };
   }
 
-  static getBalance(
-    data: GetObjectDataResponse | SuiMoveObject
-  ): BN | undefined {
+  public static getID(obj: ObjectData): ObjectId {
+    if (isSuiMoveObject(obj)) {
+      return obj.fields.id.id;
+    }
+    return getObjectId(obj);
+  }
+
+  /**
+   * Convenience method for select coin objects that has a balance greater than or equal to `amount`
+   *
+   * @param amount coin balance
+   * @param exclude object ids of the coins to exclude
+   * @return a list of coin objects that has balance greater than `amount` in an ascending order
+   */
+  static selectCoinsWithBalanceGreaterThanOrEqual(
+    coins: ObjectDataFull[],
+    amount: bigint,
+    exclude: ObjectId[] = []
+  ): ObjectDataFull[] {
+    return Coin.sortByBalance(
+      coins.filter(
+        (c) => !exclude.includes(Coin.getID(c)) && Coin.getBalance(c)! >= amount
+      )
+    );
+  }
+
+  /**
+   * Convenience method for select a minimal set of coin objects that has a balance greater than
+   * or equal to `amount`. The output can be used for `PayTransaction`
+   *
+   * @param amount coin balance
+   * @param exclude object ids of the coins to exclude
+   * @return a minimal list of coin objects that has a combined balance greater than or equal
+   * to`amount` in an ascending order. If no such set exists, an empty list is returned
+   */
+  static selectCoinSetWithCombinedBalanceGreaterThanOrEqual(
+    coins: ObjectDataFull[],
+    amount: bigint,
+    exclude: ObjectId[] = []
+  ): ObjectDataFull[] {
+    const sortedCoins = Coin.sortByBalance(
+      coins.filter((c) => !exclude.includes(Coin.getID(c)))
+    );
+
+    const total = Coin.totalBalance(sortedCoins);
+    // return empty set if the aggregate balance of all coins is smaller than amount
+    if (total < amount) {
+      return [];
+    } else if (total === amount) {
+      return sortedCoins;
+    }
+
+    let sum = BigInt(0);
+    let ret = [];
+    while (sum < total) {
+      // prefer to add a coin with smallest sufficient balance
+      const target = amount - sum;
+      const coinWithSmallestSufficientBalance = sortedCoins.find(
+        (c) => Coin.getBalance(c)! >= target
+      );
+      if (coinWithSmallestSufficientBalance) {
+        ret.push(coinWithSmallestSufficientBalance);
+        break;
+      }
+
+      const coinWithLargestBalance = sortedCoins.pop()!;
+      ret.push(coinWithLargestBalance);
+      sum += Coin.getBalance(coinWithLargestBalance)!;
+    }
+
+    return Coin.sortByBalance(ret);
```

Comment:
why do we need to sort here?

Reply:
`selectCoinsWithBalanceGreaterThanOrEqual` returns coins by ascending order. Wanted to `selectCoinSetWithCombinedBalanceGreaterThanOrEqual` also return coin by ascending order for consistency. Although we can probably just use `reverse` here as the `ret` is in descending order, but this is more clear without increasing the time complexity since on line 105 we already do sorting once